### PR TITLE
router: support downstream cert formatting in the header formatter

### LINKED ISSUE #14501: Support configurable date formats for DOWNSTREAM_PEER_CERT_V_START/END
We currently support configurable date formats using an access log command operator for START_TIME but it would be helpful to expend this behavior to DOWNSTREAM_PEER_CERT_V_START and DOWNSTREAM_PEER_CERT_V_END as well.

### PR DESCRIPTION
**Pull request by esmet**
This extends downstream cert formatting capabilities to the router's header formatter. The previous related PR only added support to the access log formatter.

The first commit implements the feature in very direct way. The second commit refactors it to reuse some existing code. I left both commits separate so that reviewer(s) can look at each implementation and decide which is better.

Commit Message: router: support downstream cert formatting in the header formatter
Additional Description: 
Risk Level: Low
Testing: Unit tests
Docs Changes: Updated custom request/response header docs
Release Notes: router: extend custom date formatting to DOWNSTREAM_PEER_CERT_V_START and DOWNSTREAM_PEER_CERT_V_END when using :ref:`custom request/response header formats <config_http_conn_man_headers_custom_request_headers>`.
Fixes #14501

### CODE REVIEW

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -10,6 +10,7 @@ Minor Behavior Changes
 *Changes that may cause incompatibilities for some users, but should not for most*
 
 * oauth filter: added the optional parameter :ref:`auth_scopes <envoy_v3_api_field_extensions.filters.http.oauth2.v3alpha.OAuth2Config.auth_scopes>` with default value of 'user' if not provided. Enables this value to be overridden in the Authorization request to the OAuth provider.
+* router: extend custom date formatting to DOWNSTREAM_PEER_CERT_V_START and DOWNSTREAM_PEER_CERT_V_END when using :ref:`custom request/response header formats <config_http_conn_man_headers_custom_request_headers>`.
```
Comment (dio) [inline docs/root/version_history/current.rst:13]: nit. extended?

File: source/common/router/header_formatter.cc [hunk 2]
```diff
@@ -331,23 +329,7 @@ StreamInfoHeaderFormatter::StreamInfoHeaderFormatter(absl::string_view field_nam
...
-    }
-    field_extractor_ = [this, pattern](const Envoy::StreamInfo::StreamInfo& stream_info) {
-      const auto& formatters = start_time_formatters_.at(pattern);
-      std::string formatted;
-      for (const auto& formatter : formatters) {
-        const auto bit = formatter->format(*Http::StaticEmptyHeaders::get().request_headers,
-                                           *Http::StaticEmptyHeaders::get().response_headers,
-                                           *Http::StaticEmptyHeaders::get().response_trailers,
-                                           stream_info, absl::string_view());
-        absl::StrAppend(&formatted, bit.value_or("-"));
```
Comment (dio) [inline source/common/router/header_formatter.cc:347]: Seems like this was not tested before?

**Review by dio — COMMENTED** [review]
While still thinking about unifying this in the access log site or here, a nit and a question:

File: source/common/router/header_formatter.cc (hunk 2 above)
Comment (esmet) [reply source/common/router/header_formatter.cc:347]: interesting, it had some tests, but none that were equivalent to the tests that check for custom formats like the ones I added for V_START/V_END. I'll update it.

Comment (esmet) [reply source/common/router/header_formatter.cc:347]: Added 👍

**Comment by esmet** [comment]
I have to force push to fix a DCO issue with 1 of the commits in this PR. I'm not totally sure how I messed it up, but I'll fix it now!

**Comment by esmet** [comment]
> I have to force push to fix a DCO issue with 1 of the commits in this PR. I'm not totally sure how I messed it up, but I'll fix it now!

Actually I see why this was messed up. It's because I made a change using the GitHub UI  to update current.rst. I will know not to do that in the future.

**Review by dio — APPROVED** [review]
Thanks!

**Comment by dio** [comment]
@zuercher do you want to take a look?

**Comment by esmet** [comment]
@zuercher friendly poke. If you think this is okay, then it should be ready to land 👍

**Review by zuercher — APPROVED** [review]

**Comment by zuercher** [comment]
Sorry for being slow.

**Pull request merged**